lib/srv/db/cassandra/engine.go

```diff
@@ -0,0 +1,308 @@
+// Cassandra wire protocol relies on streamID to that are set by the client and server response needs to
+// set the correct streamID in order to get streamID SendError reads a first message send by the client.
+func (e *Engine) SendError(sErr error) {
+	if utils.IsOKNetworkError(sErr) || sErr == nil {
+		return
+	}
+	e.Log.Debugf("Cassandra connection error: %v.", sErr)
+	if e.handshakeTriggered {
+		return
+	}
+
+	eh := failedHandshake{error: sErr}
+	if err := eh.handshake(e.clientConn, nil); err != nil {
+		e.Log.Warnf("Cassandra handshake error: %v.", sErr)
+	}
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = protocol.NewConn(clientConn)
+	e.sessionCtx = sessionCtx
+	return nil
+}
+
+// HandleConnection processes the connection from Cassandra proxy coming
+// over reverse tunnel.
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	err := e.authorizeConnection(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	serverConn, err := e.connect(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer serverConn.Close()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.handshake(sessionCtx, e.clientConn, serverConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(e.handleClientServerConn(ctx, e.clientConn, serverConn))
+}
+
+func (e *Engine) handleClientServerConn(ctx context.Context, clientConn *protocol.Conn, serverConn net.Conn) error {
+	errC := make(chan error, 2)
+	go func() {
+		err := e.handleClientConnectionWithAudit(clientConn, serverConn)
+		errC <- trace.Wrap(err, "client done")
+	}()
+	go func() {
+		err := e.handleServerConnection(serverConn)
+		errC <- trace.Wrap(err, "server done")
+	}()
+
+	var errs []error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		case err := <-errC:
+			if err != nil && !utils.IsOKNetworkError(errors.Unwrap(err)) && !errors.Is(err, io.EOF) {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return trace.NewAggregate(errs...)
+
+}
+
+func (e *Engine) handleClientConnectionWithAudit(clientConn *protocol.Conn, serverConn net.Conn) error {
+	defer serverConn.Close()
+	for {
+		packet, err := clientConn.ReadPacket()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := e.processPacket(packet); err != nil {
+			return trace.Wrap(err)
+		}
+		if _, err := serverConn.Write(packet.Raw()); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+func (e *Engine) handleServerConnection(serverConn net.Conn) error {
+	defer e.clientConn.Close()
+	if _, err := io.Copy(e.clientConn, serverConn); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func validateCassandraUsername(ses *common.Session, msg *message.AuthResponse) error {
```

Comment:
Is it checking if the user is authorized, or if Cassandra is using the same user we provided? (The function name and error are not clear).